### ISSUE TITLE
Generate Leibniz instances

### ISSUE THREAD

**Issue by jbgi** [derive4j/hkt#17]
cf. https://github.com/scalaz/scalaz/blob/series/7.3.x/core/src/main/scala/scalaz/Leibniz.scala

It will probably be useful to represent the type equality between a type and its HK encoding.
Can wait after 1.0 release.

**Comment by clinuxrulz** [comment]
One approach:
https://gist.github.com/clinuxrulz/fe9c1ffd36fb3524323e48caa846c396

`TypeEq` does the work. `ListW` is an example.

**Comment by clinuxrulz** [comment]
Leibnizian equality could also be useful in the `derive4j` library. It would allow you to encode more kinds of GADTs with complete type-safety.

**Example:**

```
@Data
public class Expr<A> {

    public interface Cases<R,A> {
        ...
        public <R> R litInt(TypeEq<Integer,A> typeEq, Integer a);
        public <R> R litString(TypeEq<String,A> typeEq, String a);
        ...
    }

    public abstract <R> R match(Cases<R> cases);

    public A eval() {
        return Exprs
            .cases()
            ...
            .litInt((typeEq, a) -> typeEq.coerce(a))
            .litString((typeEq, a) -> typeEq.coerce(a))
            ...
            .f(this);
    }

    public static Expr<Integer> litInt(int a) {
        return Exprs.litInt(TypeEq.refl(), a);
    }

    public static Expr<String> litString(String a) {
        return Exprs.litString(TypeEq.refl(), a);
    }
}
```

**Comment by jbgi** [comment]
@clinuxrulz Yes, for now derive4j use any function, eg. (Integer => A) or (String => A), as type equality witness. Using a Leibniz would be more explicit and more useful: allow type substitution with no cost and no Functor requirement.

I am not completely clear on why using a Leibniz would be more type-safe though. Do you have an example where current encoding in derive4j (https://gist.github.com/jbgi/208a1733f15cdcf78eb5) would not be type-safe?

**Comment by clinuxrulz** [comment]
Ah! Yes, the example code you gave is also completely type safe. I did not think of using just a function.

I just remember Leibniz being used to bring GADTs to Haskell98 . Must be done that way because the compiler can infer the typeclass instances.

**Comment by jbgi** [comment]
@gneuvill @clinuxrulz What do you think about adding a `Leibniz`/`TypeEq` implementation to derive4j/hkt? I think it would be a good fit. It does not need any dependency apart from `__`.

**Comment by gneuvill** [comment]
:+1:

**Comment by clinuxrulz** [comment]
sounds good.

**Comment by gneuvill** [comment]
> Using a Leibniz would be more explicit and more useful: allow type substitution with no cost and no Functor requirement.

Do you care to elaborate a bit on that ? Also, do one of you guys can point to me a _gentle_ introduction to Leibnizian equality (in the context of programming, I'm not much of a mathematician...). Thanks !

**Comment by jbgi** [comment]
> Do you care to elaborate a bit on that ? 

yes. let's take an example:

``` java
List<__<Toto.mu, A>> hkTotoList = ...
```

without Leibniz you need to use map (Functor) to recover concrete type:

``` java
List<Toto<A>> totoListViaMap = hkTotoList.map(Toto::narrow); // O(n)
```

but via a Leibniz we could do:

``` java
Leibniz<__<Toto.mu, A>, Toto<A>> totoHktEq = Toto.hktTypeEq();
List<Toto<A>> totoListViaLeibniz = List.narrow(totoHktEq.subst(hkTotoList)); // O(1)
```

> gentle introduction to Leibnizian equality (in the context of programming

I learned about Leibniz in the context of GADT via @S11001001 blog posts:
http://typelevel.org/blog/2014/07/02/type_equality_to_leibniz.html
http://typelevel.org/blog/2014/09/20/higher_leibniz.html

**Comment by gneuvill** [comment]
> but via a Leibniz we could do [...]

Hmm... Nice ! In an implementation attempt of `Traversable`, I indeed had to slaughter `traverse` and `sequence` signatures in order to avoid having to map in derivations ; Leibniz is obviously a much better solution !

> I learned about Leibniz in the context of GADT via @S11001001 blog posts

Thanks Jean-Baptiste. I'll update my readings list !

**Comment by jbgi** [comment]
Fixed by 7da75d86bfbb304a277afe2caa5481452437650e.
